icu4c/source/tools/toolutil/pkg_genc.cpp

```diff
@@ -302,29 +332,41 @@ writeAssemblyCode(const char *filename, const char *destdir, const char *optEntr
         }
     }
 
-    sprintf(bufferStr, assemblyHeader[assemblyHeaderIndex].header,
+    count = snprintf(
+        buffer.chars, sizeof(buffer.chars),
+        assemblyHeader[assemblyHeaderIndex].header,
         entry, entry, entry, entry,
         entry, entry, entry, entry);
-    T_FileStream_writeLine(out, bufferStr);
+    if (count + 1 > sizeof(buffer.chars)) {
+        fprintf(stderr, "genccode: entry name too long (long filename?)\n");
+        exit(U_ILLEGAL_ARGUMENT_ERROR);
+    }
+    T_FileStream_writeLine(out, buffer.chars);
     T_FileStream_writeLine(out, assemblyHeader[assemblyHeaderIndex].beginLine);
 
     for(;;) {
-        memset(buffer, 0, sizeof(buffer));
-        length=T_FileStream_read(in, buffer, sizeof(buffer));
+        memset(buffer.uint32s, 0, sizeof(buffer.uint32s));
+        length=T_FileStream_read(in, buffer.uint32s, sizeof(buffer.uint32s));
         if(length==0) {
             break;
         }
-        for(i=0; i<(length/sizeof(buffer[0])); i++) {
-            column = write32(out, buffer[i], column);
+        for(i=0; i<(length/sizeof(buffer.uint32s[0])); i++) {
```

Comment:
This seems fishy: What if the last read sees length not as a multiple of 4?
Leave a TODO?